### ISSUE TITLE
Log messages may be buffered indefinitely

### ISSUE THREAD

**Issue by JustinAzoff** [zeek/zeek#780]
`Manager::PublishLogWrite` contains this code:

    if ( lb.message_count >= log_batch_size ||
         (network_time - lb.last_flush >= log_batch_interval ) )
        statistics.num_logs_outgoing += lb.Flush(bstate->endpoint, log_batch_size);

Which flushes the outbound log buffers if more than `log_batch_size` message have been published or if  the  last  flush was more than `log_batch_interval` old.

If 2+ messages  are received back to back,  the first one will be logged because `network_time - lb.last_flush  >= log_batch_interval` is  true,  but the 2nd message will be buffered.  If there are no additional messages, the queue will never be flushed.  This can easily happen on a log stream that is not very active, but might see bursts of logs, like `intel` or `notice`.

Calling `Broker::__flush_logs()` periodically can fix this.

I also couldn't see where last_flush is ever initialized to something, so this might also cause weird behavior.

**Comment by jsiwek** [comment]
Thanks, yeah just having a recurring event call `Broker::flush_logs` (and removing the C++-side logic) seems simplest.  Did that in #781

**Comment by jojoob** [comment]
Thanks for fixing this.
The fix flushes all logs periodically every Broker::log_batch_interval. But when zeek is stopped pending logs will not get flushed. Or am I missing something?

**Comment by JustinAzoff** [comment]
@jojoob as part of the shutdown process anything that is pending will be flushed.  For example

src/broker/Manager.cc

```
void Manager::Terminate()
    {
    FlushLogBuffers();
    ...
```

**Comment by luguifang** [comment]
@jsiwek Is this bug fixed in 3.0.7? Why does the same problem occur when I use 3.0.7

**Comment by jsiwek** [comment]
@luguifang The patch is in 3.0.2+.  If you still have problems, create a separate issue, ideally with steps to reproduce the unexpected behavior you're seeing.
